src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+        if (elf_header.shstrndx == elf.SHN_UNDEF) return error.LongStringTable;
+
+        // First, read the string table section.
+        const string_table_shdr_offset = elf_header.shoff +
+            @sizeOf(elf.Elf64_Shdr) * elf_header.shstrndx;
+        var string_table_shdr: elf.Elf64_Shdr = undefined;
+        std.mem.copy(
+            u8,
+            std.mem.asBytes(&string_table_shdr),
+            self.source_buffer[string_table_shdr_offset..][0..@sizeOf(elf.Elf64_Shdr)],
+        );
+
+        if (string_table_shdr.sh_type != elf.SHT_STRTAB) return error.InvalidStringTable;
+        if (string_table_shdr.sh_size <= 0) return error.InvalidStringTable;
+        if (string_table_shdr.sh_size >= self.source_buffer.len) return error.InvalidStringTable;
+
+        const sh_offset = string_table_shdr.sh_offset; // Line length limits.
+        const string_table = self.source_buffer[sh_offset..][0..string_table_shdr.sh_size];
+
+        // Next, go through each ELF section to find the ones we're looking for;
+        var pack_offset: ?usize = null;
+        var metadata_offset: ?u32 = null;
+        for (0..elf_header.shnum) |i| {
+            var shdr: elf.Elf64_Shdr = undefined;
+            const offset = elf_header.shoff + @sizeOf(elf.Elf64_Shdr) * i;
+            std.mem.copy(
+                u8,
+                std.mem.asBytes(&shdr),
+                self.source_buffer[offset .. offset + @sizeOf(elf.Elf64_Shdr)],
+            );
+            const name = std.mem.sliceTo(
+                @as([*:0]const u8, @ptrCast((&string_table).ptr + shdr.sh_name)),
+                0,
+            );
+
+            if (std.mem.eql(u8, name, ".tbmvp")) {
+                // Our pack must be the second-last section in the file.
+                if (pack_offset != null) return error.MultipleMultiversionPack;
+                if (i != elf_header.shnum - 2) return error.InvalidMultiversionPackLocation;
+
+                pack_offset = @intCast(shdr.sh_offset);
+            } else if (std.mem.eql(u8, name, ".tbmvm")) {
+                // Our metadata must be the last section in the file.
+                if (metadata_offset != null) return error.MultipleMultiversionMetadata;
+                if (shdr.sh_size != @sizeOf(MultiversionMetadata))
+                    return error.InvalidMultiversionMetadataSize;
+                if (i != elf_header.shnum - 1) return error.InvalidMultiversionMetadataLocation;
+
+                metadata_offset = @intCast(shdr.sh_offset);
+            }
+        }
+
+        if (pack_offset == null or metadata_offset == null)
+            return error.MultiversionMetadataOrPackNotFound;
+
+        return .{
+            .pack = pack_offset.?,
+            .metadata = metadata_offset.?,
+        };
+    }
+
+    fn handle_error(self: *Multiversion, result: anyerror) void {
+        log_multiversioning.err("binary does not contain a valid multiversion pack: {}", .{result});
```

Comment:
Unless I'm missing something, if a bad binary is installed, this message will get printed every millisecond.

Reply:
It'll only be printed every second if the binary is missing. If the binary is there but bad and hasn't changed, the check in `on_statx` won't fire it again.